File: motor_metrics/signals.py
"""Postural primitives derived from stored pose *world* landmarks.

Everything here is recomputed from raw stored signals and nothing is written back —
the same derive-on-read rule the recording format is built on (see
``recording/recorder.py`` and ``Recording.angles()``).

**The frame.** ``landmarks_world`` is metric (meters) and **hip-centered**: MediaPipe
defines the origin *as* the midpoint of the hips. Two consequences run through this
whole package:

1. A "center of mass" proxy taken as the mid-hip in this frame is **identically zero**
   on every frame. Sway measured on it is floating-point noise. The signal that
   actually carries postural information is :func:`trunk_vector` — the mid-shoulder
   expressed relative to the pelvis, i.e. trunk-over-pelvis excursion. The hip-centered
   origin is what makes that scale-free and calibration-free, so this is a better
   signal than a true COM proxy would be, not a consolation.
2. Translation across the floor is **not recoverable** from this frame. The only
   whole-body translation signal in a recording is :func:`com_norm`, in image
   fractions. See :mod:`motor_metrics.crawl` for what that does and does not buy.

**The vertical.** There is no gravity vector in a camera-only recording. Axes are
camera-relative (x right, y down, z toward camera), so :data:`WORLD_UP` is vertical
**only if the camera is level**. That assumption is not checked automatically — it is
checked by eye, via the notebook's calibration diagnostic over a ``calib;pose=upright``
segment. Every consumer records which vertical it used (``up_source``) so a tilted
session can be found later rather than quietly biasing a year of numbers.
"""

import numpy as np
from mediapipe.tasks.python.vision import PoseLandmark as L

from pose_estimation.angles import angle_between

# Up in the camera frame: MediaPipe world coords are y-DOWN, so up is -y.
# Valid only for a level camera — see the module docstring.
WORLD_UP = np.array([0.0, -1.0, 0.0])

_EPS = 1e-9


def mid(points: np.ndarray, i: int, j: int) -> np.ndarray:
    """Midpoint of landmarks ``i`` and ``j`` per frame; ``(N, 3)``."""
    p = np.asarray(points, dtype=np.float64)
    return (p[:, i, :] + p[:, j, :]) / 2.0


def trunk_vector(world: np.ndarray) -> np.ndarray:
    """``(N, 3)`` pelvis -> mid-shoulder, in meters.

    This is the package's core postural signal, used two ways: as a **position** whose
    excursion is sway (:mod:`motor_metrics.hold`), and as a **direction** whose angle
    from vertical is lean (:func:`trunk_from_vertical`).

    The ``- mid_hip`` term is very nearly a no-op, since mid-hip *is* the world frame's
    origin. It is written explicitly anyway: it costs nothing, it makes the signal's
    meaning legible without knowing MediaPipe's origin convention, and it stays correct
    if that convention ever shifts.
    """
    w = np.asarray(world, dtype=np.float64)
    return mid(w, L.LEFT_SHOULDER, L.RIGHT_SHOULDER) - mid(w, L.LEFT_HIP, L.RIGHT_HIP)


def trunk_from_vertical(world: np.ndarray, up: np.ndarray = WORLD_UP) -> np.ndarray:
    """``(N,)`` angle in degrees between the trunk and ``up``. 0 = upright.

    Reuses :func:`pose_estimation.angles.angle_between`, so there is one definition of
    "angle" in the codebase, and inherits its **unsigned** semantics: this cannot tell
    a forward lean from a backward or lateral one. Use :func:`project_horizontal` when
    the direction of lean matters. NaN frames propagate to NaN.
    """
    up_unit = _unit(np.asarray(up, dtype=np.float64))
    trunk = trunk_vector(world)
    origin = np.zeros(3)
    # Per-frame loop mirroring Recording.angles(); N is at most a few thousand frames.
    return np.array([angle_between(t, origin, up_unit) for t in trunk])


def project_horizontal(points: np.ndarray, up: np.ndarray = WORLD_UP) -> np.ndarray:
    """Project ``(N, 3)`` points onto the plane perpendicular to ``up``; ``(N, 2)``.

    Returns ``(ML, AP)`` — medio-lateral and antero-posterior — and the axis order is
    the point of the function. On a single camera, ML lies in the image plane and is
    measured well, while AP is MediaPipe's inferred depth and is markedly noisier. A
    caller that collapses these into one isotropic sway number averages a good estimate
    with a bad one and cannot tell afterwards which it was looking at.

    For a level camera this is exactly ``ML = world x``, ``AP = world z``. For a tilted
    ``up`` the basis is the closest thing to that which is still orthonormal.
    """
    p = np.asarray(points, dtype=np.float64)
    up_unit = _unit(np.asarray(up, dtype=np.float64))

    # ML is world-x with any `up` component removed, so it stays the image-plane axis.
    ml = _reject(np.array([1.0, 0.0, 0.0]), up_unit)
    if np.linalg.norm(ml) < 1e-6:
        # Degenerate: `up` is (nearly) world-x, i.e. the camera is rolled ~90 degrees.
        # Fall back to world-z so the basis stays defined rather than blowing up.
        ml = _reject(np.array([0.0, 0.0, 1.0]), up_unit)
    ml = _unit(ml)
    ap = _unit(np.cross(up_unit, ml))
    return np.stack([p @ ml, p @ ap], axis=-1)


def com_norm(norm: np.ndarray) -> np.ndarray:
    """``(N, 3)`` mid-hip in **normalized image coordinates**, not meters.

    The only whole-body translation signal a recording has: ``landmarks_world`` is
    hip-centered, so the pelvis cannot move in it by construction. Columns 0 and 1 are
    image fractions of width and height; column 2 is MediaPipe's relative depth, which
    is on a different and much weaker scale — do not mix it with the first two.

    Any speed derived from this is in image-widths per second. Converting it to m/s
    needs a scale reference; see :func:`motor_metrics.crawl.scale_m_per_norm` for what
    that costs in accuracy.
    """
    return mid(np.asarray(norm, dtype=np.float64), L.LEFT_HIP, L.RIGHT_HIP)


def estimate_up(rec, seg) -> np.ndarray:
    """Unit "up" from a ``calib;pose=upright`` segment: the median trunk direction.

    **Opt-in, and deliberately not the default.** It cannot separate "the camera is
    tilted" from "the child's trunk is not actually vertical" — and for a child with
    global developmental delay the second term is not small, so calibrating on it can
    bake a real postural asymmetry into the reference and hide it forever after.

    Use it when you know the camera was tilted and you trust the calibration pose. Pass
    the result as ``up=`` and the consuming metric will record ``up_source="calib"``.
    """
    trunk = trunk_vector(rec.landmarks_world)[seg.start : seg.stop]
    good = trunk[~np.isnan(trunk).any(axis=1)]
    if good.size == 0:
        raise ValueError("Calibration segment has no frames with a detected pose.")
    return _unit(np.median(good, axis=0))


def _unit(v: np.ndarray) -> np.ndarray:
    n = np.linalg.norm(v)
    return v / n if n > _EPS else v


def _reject(v: np.ndarray, unit_axis: np.ndarray) -> np.ndarray:
    """Component of ``v`` perpendicular to the unit vector ``unit_axis``."""
    return v - np.dot(v, unit_axis) * unit_axis

File: android/metrics/src/main/kotlin/dev/remapy/metrics/Signals.kt
package dev.remapy.metrics

import kotlin.math.sqrt

/**
 * Postural primitives derived from pose *world* landmarks.
 *
 * Port of `motor_metrics/signals.py`, whose docstring carries the reasoning. Two facts from
 * it run through everything downstream and are worth restating where the code is:
 *
 * **The frame is hip-centered.** MediaPipe puts the world origin *at* the midpoint of the
 * hips. So (1) a mid-hip "centre of mass" proxy is identically zero on every frame and sway
 * measured on it is floating-point noise — the signal that carries postural information is
 * [trunkVector], the mid-shoulder relative to the pelvis, and the hip-centered origin is what
 * makes it scale- and calibration-free; (2) translation across the floor is **not
 * recoverable**, and the only whole-body translation signal is [comNorm], in image fractions.
 *
 * **There is no gravity vector in a camera-only recording.** Axes are camera-relative, so
 * [WORLD_UP] is vertical **only if the camera is level** — a real risk on a hand-held or
 * propped phone, and the reason the live hold readout leads with the trunk angle's delta from
 * its own window baseline rather than an absolute lean. Every consumer records which vertical
 * it used, so a tilted session can be found later rather than quietly biasing a year of numbers.
 */
object Signals {

    /** Up in the camera frame: MediaPipe world coords are y-DOWN, so up is -y. */
    val WORLD_UP = doubleArrayOf(0.0, -1.0, 0.0)

    private const val EPS = 1e-9

    /** Midpoint of landmarks [i] and [j] for one frame. */
    fun mid(worldRow: FloatArray, i: Int, j: Int): DoubleArray {
        val a = worldRow.point(i)
        val b = worldRow.point(j)
        return DoubleArray(3) { (a[it] + b[it]) / 2.0 }
    }

    /**
     * Pelvis -> mid-shoulder, in meters. The package's core postural signal.
     *
     * Used two ways: as a **position** whose excursion is sway, and as a **direction** whose
     * angle from vertical is lean. The `- midHip` term is very nearly a no-op since mid-hip
     * *is* the origin; it is written anyway because it costs nothing and makes the signal's
     * meaning legible without knowing MediaPipe's convention.
     */
    fun trunkVector(worldRow: FloatArray): DoubleArray {
        val shoulders = mid(worldRow, Landmarks.LEFT_SHOULDER, Landmarks.RIGHT_SHOULDER)
        val hips = mid(worldRow, Landmarks.LEFT_HIP, Landmarks.RIGHT_HIP)
        return DoubleArray(3) { shoulders[it] - hips[it] }
    }

    /** [trunkVector] over a span of frames, as an `(N, 3)` matrix. */
    fun trunkVectors(frames: PoseFrames, span: Span): Matrix {
        val out = Matrix(span.nFrames, 3)
        for (i in 0 until span.nFrames) {
            val t = trunkVector(frames.worldRow(span.start + i))
            for (c in 0 until 3) out[i, c] = t[c]
        }
        return out
    }

    /**
     * Angle in degrees between the trunk and [up]; 0 = upright.
     *
     * Reuses [Angles.angleBetween] so there is one definition of "angle" in the codebase, and
     * inherits its **unsigned** semantics: this cannot tell a forward lean from a backward or
     * a lateral one. Use [projectHorizontal] when the direction of lean matters. NaN frames
     * propagate to NaN.
     */
    fun trunkFromVertical(frames: PoseFrames, span: Span, up: DoubleArray = WORLD_UP): DoubleArray {
        val upUnit = unit(up)
        val origin = doubleArrayOf(0.0, 0.0, 0.0)
        return DoubleArray(span.nFrames) {
            Angles.angleBetween(trunkVector(frames.worldRow(span.start + it)), origin, upUnit)
        }
    }

    /**
     * Project `(N, 3)` points onto the plane perpendicular to [up], giving `(ML, AP)`.
     *
     * **The axis order is the point of the function.** On a single camera, ML lies in the
     * image plane and is measured well, while AP is MediaPipe's inferred depth and is
     * markedly noisier. A caller that collapses these into one isotropic sway number averages
     * a good estimate with a bad one and cannot tell afterwards which it was looking at.
     *
     * For a level camera this is exactly `ML = world x`, `AP = world z`. For a tilted [up] the
     * basis is the closest thing to that which is still orthonormal.
     */
    fun projectHorizontal(points: Matrix, up: DoubleArray = WORLD_UP): Matrix {
        val upUnit = unit(up)
        // ML is world-x with any `up` component removed, so it stays the image-plane axis.
        var ml = reject(doubleArrayOf(1.0, 0.0, 0.0), upUnit)
        if (norm(ml) < 1e-6) {
            // Degenerate: `up` is (nearly) world-x, i.e. the camera is rolled ~90 degrees.
            // Fall back to world-z so the basis stays defined rather than blowing up.
            ml = reject(doubleArrayOf(0.0, 0.0, 1.0), upUnit)
        }
        ml = unit(ml)
        val ap = unit(cross(upUnit, ml))

        val out = Matrix(points.rows, 2)
        for (i in 0 until points.rows) {
            var a = 0.0
            var b = 0.0
            for (c in 0 until 3) {
                a += points[i, c] * ml[c]
                b += points[i, c] * ap[c]
            }
            out[i, 0] = a
            out[i, 1] = b
        }
        return out
    }

    /**
     * Mid-hip in **normalized image coordinates**, not meters; columns 0 and 1 only.
     *
     * The only whole-body translation signal a recording has, since `landmarks_world` is
     * hip-centered and the pelvis cannot move in it by construction. Column 2 of the source is
     * MediaPipe's relative depth, on a different and much weaker scale, and is deliberately
     * dropped here rather than mixed in.
     *
     * Any speed derived from this is in **image widths per second**, and it is never called
     * metres. Converting it honestly needs a scale reference this pipeline does not have.
     */
    fun comNorm(frames: PoseFrames, span: Span): Matrix {
        val out = Matrix(span.nFrames, 2)
        for (i in 0 until span.nFrames) {
            val m = mid(frames.normRow(span.start + i), Landmarks.LEFT_HIP, Landmarks.RIGHT_HIP)
            out[i, 0] = m[0]
            out[i, 1] = m[1]
        }
        return out
    }

    internal fun norm(v: DoubleArray): Double {
        var acc = 0.0
        for (x in v) acc += x * x
        return sqrt(acc)
    }

    internal fun unit(v: DoubleArray): DoubleArray {
        val n = norm(v)
        return if (n > EPS) DoubleArray(v.size) { v[it] / n } else v.copyOf()
    }

    /** Component of [v] perpendicular to the unit vector [axis]. */
    private fun reject(v: DoubleArray, axis: DoubleArray): DoubleArray {
        var dot = 0.0
        for (i in v.indices) dot += v[i] * axis[i]
        return DoubleArray(v.size) { v[it] - dot * axis[it] }
    }

    private fun cross(a: DoubleArray, b: DoubleArray): DoubleArray = doubleArrayOf(
        a[1] * b[2] - a[2] * b[1],
        a[2] * b[0] - a[0] * b[2],
        a[0] * b[1] - a[1] * b[0],
    )
}
